File: ios/Sources/AgentlyAppFoundation/Support/ForgeAgentlyDataSourceLoader.swift
import Foundation
import AgentlySDK
import ForgeIOSRuntime

func makeForgeAgentlyDataSourceLoader(
    client: AgentlyClient,
    conversationIDProvider: @escaping @Sendable () async -> String? = { nil }
) -> @Sendable (ForgeRuntime.DataSourceFetchRequest) async throws -> ForgeRuntime.DataSourceFetchResult? {
    return { request in
        let service = request.dataSource.service
        let endpoint = service?.endpoint?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard endpoint == "agentlyAPI" else {
            return nil
        }
        let uri = service?.uri?.trimmingCharacters(in: .whitespacesAndNewlines)
            ?? request.dataSource.uri?.trimmingCharacters(in: .whitespacesAndNewlines)
            ?? ""
        let normalizedURI = uri.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if normalizedURI == "v1/api/agently/scheduler" {
            return try await loadSchedules(client: client, request: request)
        }
        if normalizedURI == "v1/api/agently/scheduler/run" {
            return try await loadScheduleRuns(client: client, request: request)
        }
        if normalizedURI == "v1/workspace/metadata" {
            return try await loadWorkspaceOptions(client: client, request: request)
        }
        if normalizedURI == "v1/workspace/metadata/publicagents" {
            return try await loadPublicAgents(client: client)
        }
        guard let datasourceID = extractDatasourceID(from: uri) else {
            return nil
        }

        var inputs = request.resolvedInputs
        if let nestedInput = request.input.parameters["input"] {
            inputs["input"] = nestedInput
        }
        if let page = request.input.parameters["page"] {
            inputs["page"] = page
        }
        for (key, value) in request.input.parameters where key != "input" && key != "page" && key != "parameters" {
            if inputs[key] == nil {
                inputs[key] = value
            }
        }
        if !request.input.filter.isEmpty {
            if inputs["input"]?.objectValue != nil {
                var inputObject = inputs["input"]?.objectValue ?? [:]
                var queryObject = inputObject["query"]?.objectValue ?? [:]
                queryObject.merge(request.input.filter) { _, new in new }
                inputObject["query"] = .object(queryObject)
                inputs["input"] = .object(inputObject)
            } else {
                inputs.merge(request.input.filter) { _, new in new }
            }
        }
        if let page = request.input.page {
            inputs["page"] = .number(Double(page))
        }
        if let paging = request.dataSource.paging, paging.enabled != false {
            let pageKey = paging.parameters["page"]?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            let sizeKey = paging.parameters["size"]?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            if !pageKey.isEmpty, inputs[pageKey] == nil {
                inputs[pageKey] = .number(Double(request.input.page ?? 1))
            }
            if !sizeKey.isEmpty, inputs[sizeKey] == nil, let size = paging.size, size > 0 {
                inputs[sizeKey] = .number(Double(size))
            }
        }

        let requestConversationID = request.conversationID?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let fallbackConversationID = requestConversationID.isEmpty
            ? await conversationIDProvider()?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            : ""
        let conversationID = requestConversationID.isEmpty ? fallbackConversationID : requestConversationID
        let response = try await client.fetchDatasource(
            FetchDatasourceInput(
                id: datasourceID,
                inputs: inputs.isEmpty ? nil : inputs.mapValues(\.appValue),
                conversationId: conversationID.isEmpty ? nil : conversationID
            )
        )
        return ForgeRuntime.DataSourceFetchResult(
            rows: response.rows.map { $0.mapValues(\.forgeValue) },
            metrics: response.metrics?.mapValues(\.forgeValue) ?? [:]
        )
    }
}

private func loadSchedules(
    client: AgentlyClient,
    request: ForgeRuntime.DataSourceFetchRequest
) async throws -> ForgeRuntime.DataSourceFetchResult {
    let query = request.input.filter["name"]?.stringValue?
        .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    let schedules = try await client.listSchedules().filter {
        query.isEmpty || $0.name.localizedCaseInsensitiveContains(query)
    }
    let pageSize = max(1, request.dataSource.paging?.size ?? max(1, schedules.count))
    let page = max(1, request.input.page ?? 1)
    let start = min(schedules.count, (page - 1) * pageSize)
    let rows = try schedules.dropFirst(start).prefix(pageSize).map(scheduleForgeRow)
    let pageCount = schedules.isEmpty ? 0 : (schedules.count + pageSize - 1) / pageSize
    return ForgeRuntime.DataSourceFetchResult(
        rows: rows,
        metrics: [
            "pageCount": .number(Double(pageCount)),
            "totalCount": .number(Double(schedules.count))
        ]
    )
}

private func scheduleForgeRow(_ schedule: Schedule) throws -> [String: ForgeIOSRuntime.JSONValue] {
    let data = try JSONEncoder.agently().encode(schedule)
    let value = try JSONDecoder.agently().decode(AgentlySDK.JSONValue.self, from: data)
    return value.forgeValue.objectValue ?? [:]
}

private func loadWorkspaceOptions(
    client: AgentlyClient,
    request: ForgeRuntime.DataSourceFetchRequest
) async throws -> ForgeRuntime.DataSourceFetchResult {
    let metadata = try await client.getWorkspaceMetadata()
    let rows: [[String: ForgeIOSRuntime.JSONValue]]
    switch request.dataSource.selectors?.data?.trimmingCharacters(in: .whitespacesAndNewlines) {
    case "agentInfos":
        rows = metadata.agentInfos.map {
            [
                "id": .string($0.agentID ?? $0.id),
                "name": .string($0.name ?? $0.agentID ?? "Agent"),
                "modelRef": .string($0.modelRef ?? "")
            ]
        }
    case "modelInfos":
        let models = metadata.modelInfos.map {
            [
                "id": ForgeIOSRuntime.JSONValue.string($0.modelID ?? $0.id),
                "name": .string($0.name ?? $0.modelID ?? "Model")
            ]
        }
        rows = models.isEmpty
            ? metadata.models.map { ["id": .string($0), "name": .string($0)] }
            : models
    default:
        rows = []
    }
    return ForgeRuntime.DataSourceFetchResult(rows: rows)
}

private func loadPublicAgents(client: AgentlyClient) async throws -> ForgeRuntime.DataSourceFetchResult {
    let agents = try await client.getPublicAgents()
    return ForgeRuntime.DataSourceFetchResult(rows: agents.map {
        [
            "id": .string($0.agentID ?? $0.id),
            "name": .string($0.name ?? $0.agentID ?? "Agent"),
            "modelRef": .string($0.modelRef ?? "")
        ]
    })
}

private func loadScheduleRuns(
    client: AgentlyClient,
    request: ForgeRuntime.DataSourceFetchRequest
) async throws -> ForgeRuntime.DataSourceFetchResult {
    let filters = request.input.filter.reduce(into: [String: String]()) { result, entry in
        if let value = entry.value.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty {
            result[entry.key] = value
        }
    }
    let scheduleID = request.resolvedInputs["scheduleId"]?.stringValue
        ?? request.input.parameters["scheduleId"]?.stringValue
    let page = max(1, request.input.page ?? 1)
    let size = max(1, request.dataSource.paging?.size ?? 10)
    let response = try await client.listScheduleRuns(
        scheduleID: scheduleID,
        filters: filters,
        page: page,
        size: size
    )
    let rows = try response.rows.map { run -> [String: ForgeIOSRuntime.JSONValue] in
        let data = try JSONEncoder.agently().encode(run)
        let value = try JSONDecoder.agently().decode(AgentlySDK.JSONValue.self, from: data)
        return value.forgeValue.objectValue ?? [:]
    }
    return ForgeRuntime.DataSourceFetchResult(
        rows: rows,
        metrics: [
            "pageCount": .number(Double(response.pageCount)),
            "totalCount": .number(Double(response.totalCount))
        ]
    )
}

internal func extractDatasourceID(from uri: String) -> String? {
    let trimmed = uri.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return nil }
    let marker = "/datasources/"
    guard let markerRange = trimmed.range(of: marker) else { return nil }
    var suffix = String(trimmed[markerRange.upperBound...])
    if let queryRange = suffix.range(of: "?") {
        suffix = String(suffix[..<queryRange.lowerBound])
    }
    if suffix.hasSuffix("/fetch") {
        suffix.removeLast("/fetch".count)
    }
    let datasourceID = suffix.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    guard !datasourceID.isEmpty else { return nil }
    return datasourceID.removingPercentEncoding ?? datasourceID
}

File: android/app/src/main/java/com/viant/agently/android/ForgeAgentlyDataSourceLoader.kt
package com.viant.agently.android

import android.util.Log
import com.viant.agentlysdk.AgentlyClient
import com.viant.agentlysdk.FetchDatasourceInput
import com.viant.agentlysdk.Schedule
import com.viant.agentlysdk.ScheduleRun
import com.viant.agentlysdk.fetchDatasource
import com.viant.forgeandroid.runtime.ForgeRuntime
import com.viant.forgeandroid.runtime.JsonUtil

internal fun makeForgeAgentlyDataSourceLoader(
    client: AgentlyClient
): suspend (ForgeRuntime.DataSourceFetchRequest) -> ForgeRuntime.DataSourceFetchResult? {
    return loader@{ request ->
        val startedAt = System.currentTimeMillis()
        val service = request.dataSource.service
        val endpoint = service?.endpoint?.trim().orEmpty()
        val uri = service?.uri?.trim().orEmpty()

        if (endpoint.isNotEmpty() && !endpoint.equals("agentlyAPI", ignoreCase = true)) {
            return@loader null
        }
        if (uri.trimEnd('/') == "/v1/api/agently/scheduler") {
            return@loader loadSchedules(client, request)
        }
        if (uri.trimEnd('/') == "/v1/api/agently/scheduler/run") {
            return@loader loadScheduleRuns(client, request)
        }
        if (uri.trimEnd('/') == "/v1/workspace/metadata") {
            return@loader loadWorkspaceOptions(client, request)
        }
        if (uri.trimEnd('/') == "/v1/workspace/metadata/publicagents") {
            return@loader loadPublicAgents(client)
        }
        val datasourceId = extractDatasourceId(uri) ?: return@loader null

        val inputs = request.resolvedInputs.toMutableMap()
        val rawParameters = request.input.parameters
        val nestedInput = rawParameters["input"]
        if (nestedInput != null) {
            inputs["input"] = nestedInput
        }
        rawParameters["page"]?.let { inputs["page"] = it }
        rawParameters
            .filterKeys { it !in setOf("input", "page", "parameters") }
            .forEach { (key, value) ->
                if (inputs[key] == null) {
                    inputs[key] = value
                }
            }
        if (request.input.filter.isNotEmpty()) {
            if (inputs["input"] is Map<*, *>) {
                val inputObject = JsonUtil.asStringMap(inputs["input"]).toMutableMap()
                val queryObject = JsonUtil.asStringMap(inputObject["query"]).toMutableMap()
                queryObject.putAll(request.input.filter)
                inputObject["query"] = queryObject
                inputs["input"] = inputObject
            } else {
                inputs.putAll(request.input.filter)
            }
        }
        request.input.page?.let { inputs["page"] = it }
        request.dataSource.paging?.takeIf { it.enabled != false }?.let { paging ->
            val pageKey = paging.parameters["page"]?.trim().orEmpty()
            val sizeKey = paging.parameters["size"]?.trim().orEmpty()
            if (pageKey.isNotEmpty() && inputs[pageKey] == null) {
                inputs[pageKey] = request.input.page ?: 1
            }
            val size = paging.size
            if (sizeKey.isNotEmpty() && inputs[sizeKey] == null && size != null && size > 0) {
                inputs[sizeKey] = size
            }
        }

        val response = try {
            if (BuildConfig.DEBUG) {
                runCatching {
                    Log.d(
                        "ForgeDataSource",
                        "fetch start id=$datasourceId conversationId=${request.conversationId.orEmpty()} inputs=${JsonUtil.anyToElement(inputs)}"
                    )
                }
            }
            client.fetchDatasource(
                FetchDatasourceInput(
                    id = datasourceId,
                    inputs = inputs.takeIf { it.isNotEmpty() }?.mapValues { JsonUtil.anyToElement(it.value) },
                    conversationId = request.conversationId?.takeIf { it.isNotBlank() }
                )
            )
        } catch (err: Throwable) {
            // android.util.Log is a throwing stub in local JVM tests. Logging
            // must never replace the real datasource exception being tested.
            runCatching {
                Log.e(
                    "ForgeDataSource",
                    "fetch failed id=$datasourceId conversationId=${request.conversationId.orEmpty()}",
                    err
                )
            }
            throw err
        }
        if (BuildConfig.DEBUG) {
            runCatching {
                Log.d(
                    "ForgeDataSource",
                    "fetch completed id=$datasourceId rows=${response.rows.size} elapsedMs=${System.currentTimeMillis() - startedAt}"
                )
            }
        }
        ForgeRuntime.DataSourceFetchResult(
            rows = response.rows.map { row -> row.mapValues { JsonUtil.elementToAny(it.value) } },
            metrics = response.metrics?.mapValues { JsonUtil.elementToAny(it.value) } ?: emptyMap()
        )
    }
}

private suspend fun loadPublicAgents(client: AgentlyClient): ForgeRuntime.DataSourceFetchResult {
    val rows = client.getPublicAgents().map { info ->
        mapOf("id" to info.id, "name" to info.name, "modelRef" to info.modelRef)
    }
    return ForgeRuntime.DataSourceFetchResult(rows = rows)
}

private suspend fun loadScheduleRuns(
    client: AgentlyClient,
    request: ForgeRuntime.DataSourceFetchRequest
): ForgeRuntime.DataSourceFetchResult {
    val page = (request.input.page ?: 1).coerceAtLeast(1)
    val size = request.dataSource.paging?.size?.takeIf { it > 0 } ?: 10
    val filters = request.input.filter.mapNotNull { (key, value) ->
        value?.toString()?.let { key to it }
    }.toMap()
    val scheduleId = request.resolvedInputs["scheduleId"]?.toString()
        ?: request.input.parameters["scheduleId"]?.toString()
    val response = client.listScheduleRuns(scheduleId, filters, page, size)
    return ForgeRuntime.DataSourceFetchResult(
        rows = response.rows.map(::scheduleRunRow),
        metrics = mapOf("pageCount" to response.pageCount, "totalCount" to response.totalCount)
    )
}

private suspend fun loadSchedules(
    client: AgentlyClient,
    request: ForgeRuntime.DataSourceFetchRequest
): ForgeRuntime.DataSourceFetchResult {
    val query = request.input.filter["name"]?.toString()?.trim().orEmpty()
    val all = client.listSchedules()
        .filter { query.isBlank() || it.name.contains(query, ignoreCase = true) }
    val pageSize = request.dataSource.paging?.size?.takeIf { it > 0 } ?: all.size.coerceAtLeast(1)
    val page = (request.input.page ?: 1).coerceAtLeast(1)
    val from = ((page - 1) * pageSize).coerceAtMost(all.size)
    val rows = all.drop(from).take(pageSize).map(::scheduleRow)
    val pageCount = if (all.isEmpty()) 0 else (all.size + pageSize - 1) / pageSize
    return ForgeRuntime.DataSourceFetchResult(
        rows = rows,
        metrics = mapOf("pageCount" to pageCount, "totalCount" to all.size)
    )
}

private suspend fun loadWorkspaceOptions(
    client: AgentlyClient,
    request: ForgeRuntime.DataSourceFetchRequest
): ForgeRuntime.DataSourceFetchResult {
    val metadata = client.getWorkspaceMetadata()
    val rows = when (request.dataSource.selectors?.data?.trim()) {
        "agentInfos" -> metadata.agentInfos.map { info ->
            mapOf("id" to info.id, "name" to info.name, "modelRef" to info.modelRef)
        }
        "modelInfos" -> metadata.modelInfos.map { info ->
            mapOf("id" to info.id, "name" to info.name)
        }.ifEmpty {
            metadata.models.map { id -> mapOf("id" to id, "name" to id) }
        }
        else -> emptyList()
    }
    return ForgeRuntime.DataSourceFetchResult(rows = rows)
}

private fun scheduleRow(schedule: Schedule): Map<String, Any?> = mapOf(
    "id" to schedule.id,
    "name" to schedule.name,
    "description" to schedule.description,
    "visibility" to schedule.visibility,
    "agentRef" to schedule.agentRef,
    "modelOverride" to schedule.modelOverride,
    "userCredUrl" to schedule.userCredUrl,
    "enabled" to schedule.enabled,
    "startAt" to schedule.startAt,
    "endAt" to schedule.endAt,
    "scheduleType" to schedule.scheduleType,
    "cronExpr" to schedule.cronExpr,
    "intervalSeconds" to schedule.intervalSeconds,
    "timezone" to schedule.timezone,
    "timeoutSeconds" to schedule.timeoutSeconds,
    "taskPromptUri" to schedule.taskPromptUri,
    "taskPrompt" to schedule.taskPrompt,
    "nextRunAt" to schedule.nextRunAt,
    "lastRunAt" to schedule.lastRunAt,
    "lastStatus" to schedule.lastStatus,
    "lastError" to schedule.lastError,
    "createdAt" to schedule.createdAt,
    "updatedAt" to schedule.updatedAt
)

private fun scheduleRunRow(run: ScheduleRun): Map<String, Any?> = mapOf(
    "id" to run.id,
    "scheduleId" to run.scheduleId,
    "conversationId" to run.conversationId,
    "status" to run.status,
    "errorMessage" to run.errorMessage,
    "scheduledFor" to run.scheduledFor,
    "startedAt" to run.startedAt,
    "completedAt" to run.completedAt,
    "createdAt" to run.createdAt,
    "updatedAt" to run.updatedAt
)

private fun extractDatasourceId(uri: String): String? {
    val normalized = uri.trim().substringBefore('?')
    val marker = "/v1/api/datasources/"
    val start = normalized.indexOf(marker)
    if (start == -1) {
        return null
    }
    val suffix = normalized.substring(start + marker.length)
    val id = suffix.substringBefore("/fetch").trim().trim('/')
    return id.ifBlank { null }
}
